### ISSUE TITLE
segmentation fault using DeDRM v7.2.1 with Calibre

### ISSUE THREAD

**Issue by dhdurgee** [noDRM/DeDRM_tools#79]
I am attempting to add one of my epub books to Calibre and encountering problems. Looking in the documentation here I attempted to do so in a terminal window with the following results:

```
dhdurgee@z560:~/Downloads$ calibredb add ./Created\ Equal.epub --with-library=~/Calibre\ Library
DeDRM v7.2.1: Trying to decrypt Created Equal.epub
DeDRM v7.2.1: Verifying zip archive integrity
DeDRM v7.2.1: Created Equal.epub is a secure Adobe Adept ePub
DeDRM v7.2.1: Trying Encryption key RTC
Segmentation fault (core dumped)
dhdurgee@z560:~/Downloads$ 
````

Running Calibre 5.42 on Linux mint x64 20.3 here.

Where do I go from here in getting this problem corrected?

Dave

**Comment by ElleKayEm** [comment]
The current version here is DeDRM 10.0.2.

**Comment by dhdurgee** [comment]
Do I simply install your 10.0.2 version over the 7.2.1 version, or are some configuration file changes necessary?

Is it sufficient in case it is totally compatible to simply replace ~/.config/calibre/plugins/DeDRM.zip with your version?

**Comment by ElleKayEm** [comment]
Should be fine to install 10.0.2 over 7.2.1.

**Comment by dhdurgee** [comment]
Installed your release over 7.2.1 here and tried again:
```
dhdurgee@z560:~/Downloads$ calibredb add ./Created\ Equal.epub --with-library=~/Calibre\ Library
DeDRM v10.0.2: Trying to decrypt Created Equal.epub
DeDRM v10.0.2: Verifying zip archive integrity
DeDRM v10.0.2: Created Equal.epub is a secure Adobe Adept ePub for UUID 500ec903-efb8-45ec-8cc0-8ff5db93808a
DeDRM v10.0.2: Trying Encryption key RTC
Segmentation fault (core dumped)
dhdurgee@z560:~/Downloads$ 
```

So where do we go from here?

**Comment by dhdurgee** [comment]
I have used the plugin successfully in the past, but of course that is with earlier versions.  I mostly get my books in Kindle format, so it is relatively rare that I need to address this case.  Once I have the book added to my Calibre library I don't keep the encrypted version around, so I don't have any others to test with.

Can I use calibre-debug to get further information on this?  I saw the documentation on how to use the command line to do so, but not how to use calibre-debug.  Hopefully seeing which code is crashing and seeing how it is being called will clarify what is going on.

**Comment by dhdurgee** [comment]
Any progress?

**Comment by noDRM** [comment]
I responded to your last email five days ago, did that get lost in a spam filter?

The book you used is perfectly valid, but your encryption key file ("RTC.der") isn't the actual account key but some other certificate instead. Adobe key files are usually between 600 and 620 bytes in size, while yours is over 1000 bytes; and the format also looks different. OpenSSL probably doesn't validate the file, just tries to decrypt stuff, and then crashes. 

Where does this key come from? Did my plugin automatically import that (unlikely) or did you manually import it from elsewhere? 

Try either exporting a new copy of your account key from ADE, or maybe try using the ACSM Input plugin (where the DeDRM plugin should automatically import Adobe keys from).

**Comment by dhdurgee** [comment]
That key has been in my configuration for a long time, along with another key names "default" and this is the first time I remember hitting a segfault.  I removed this key and attempted to add the book from the GUI.  This succeeded, so apparently "default" was the proper key to use.

I don't remember where the bad  key came from at this point.  I do have a few Barnes & Noble books, perhaps it actually belongs there as opposed to ADE.

If it would help you prevent the segfaults in the future I will provide you a copy of the proper key so you can debug the loop that is supposed to try the various keys in the ADE configuration.  I would assume the proper behavior would be to issue a message about the key being incorrect for the book and going on to the next key in the list.

Thank you for your assistance in debugging this problem.

**Comment by noDRM** [comment]
The plugin already has a bunch of code to automatically try all available keys; and when I imported your RTC key into my Calibre setup, that's exactly what it did - it attempted all keys and then complained that none of them could decrypt the book. In my case, OpenSSL even complained that "RTC" isn't even a valid key. 

I have no idea why that didn't happen in your case - Linux Mint 20.3, according to a quick Google search, seems to be based on Ubuntu 20.04 which is what I used for this test, so I'm not sure where that came from. 

As this is a segfault, I also can't just add a try/except block around the decryption code considering the crash happens in native code. 
Only thing I could try to do would be to check all the keys manually with some heuristics (how likely is it that a given key is a valid Adobe book key) and then try the key first that look like actual keys, but I'm not sure if this is worth it, considering you're the first to run into this issue in all the years, as far as I know.

**Comment by ElleKayEm** [comment]
Are you using the official Linux calibre or one that came with your Linux distribution?  I don't know if that could be a problem here, but I have seen it be an issue before.

**Comment by dhdurgee** [comment]
Yes, I am using the current official release.  As I noted earlier, there appears to be an update to OpenSSL that is causing issues with the ASCM plugin.  Perhaps that is the reason I am hitting a segfault when you simply get an error message.

**Comment by noDRM** [comment]
That large OpenSSL update is not included in Ubuntu 20.04, it's only been added in 22.04, so the issue in 20.04 should be unrelated to this.
